### ISSUE TITLE
special token loss computation

### ISSUE THREAD

**Issue by disperaller** [yongzhuo/LLaMA3-SFT#1]
![image](https://github.com/yongzhuo/LLaMA3-SFT/assets/13672297/a5810cf6-30b3-4e50-adca-5a37c85300e6)

这个地方这样写的话 那些special token就都不参与loss计算了吧？
按照这个地方的提及，貌似instruct训练的时候是需要训练那些special token的
如果不训练的话 会不会对模型产出这些special token的能力产生影响？

![image](https://github.com/yongzhuo/LLaMA3-SFT/assets/13672297/9c66797d-d6c3-458e-b5ed-21a642e39c66)

**Comment by yongzhuo** [comment]
这个是SFT只计算y的loss，下面else有x+y全部计算的选择，只需要设置 use_all_loss=True就可以了。

**Comment by disperaller** [comment]
> 这个是SFT只计算y的loss，下面else有x+y全部计算的选择，只需要设置 use_all_loss=True就可以了。

另外 我想问下 这个地方的mask 是不是1和0反掉了 怎么padding反而mask为True 真实的token反而mask为False了？

**Comment by yongzhuo** [comment]
你是对的，应该是我弄错了。
```
参数说明
attention_mask (`torch.Tensor` of shape `(batch_size, sequence_length)`, *optional*):
            Mask to avoid performing attention on padding token indices. Mask values selected in `[0, 1]`:

            - 1 for tokens that are **not masked**,
            - 0 for tokens that are **masked**.

            [What are attention masks?](../glossary#attention-mask)
```

```
实现逻辑
# Since attention_mask is 1.0 for positions we want to attend and 0.0 for
# masked positions, this operation will create a tensor which is 0.0 for
# positions we want to attend and the dtype's smallest value for masked positions.
# Since we are adding it to the raw scores before the softmax, this is
# effectively the same as removing these entirely.
extended_attention_mask = extended_attention_mask.to(dtype=dtype)  # fp16 compatibility
extended_attention_mask = (1.0 - extended_attention_mask) * torch.finfo(dtype).min
return extended_attention_mask
```
